### PR TITLE
Upgrade bitstyles to 5.0.1

### PR DESCRIPTION
**Pull request by angelikatyborska**
Resolves https://github.com/bitcrowd/bitstyles_phoenix/issues/122

Note that the sidebar example in 4.3.0 will look weird because it contains a lot of example content that's written with attributes supported in 5.0.0 only.

Also note that the class `a-button` is now `display: flex` instead of `display: inline-flex` and now `<a>`s with that class look different than `<button>`s with that class due to different default widths of the two elements. This discrepancy would need to be fixed on a project level where bitstyles-phoenix is used, by something like `a { width: fit-content }` or so. Or just be using those elements inside appropriate flex/grid containers.

List of changes:
- Button rewrite (see linked issue)
- `u-gap-m` is `u-gap-l`, looks identical
-  `a-list-reset` is `u-list-none`
- `a-badge--gray` is `a-badge--text`
- Brand2 color changed from yellow to blue, nothing to do about that
- Color palette changed a lot, causing `u-fg-` and `u-bg-` classes to be different. Darren helped me choose adequate replacements.
- Size naming changed to use a number instead of repeating a letter, e.g. xxxs is 3xs.

### CODE REVIEW

**Comment by angelikatyborska** [comment]
Waiting for bitstyles `v5.0.0` to be republished as `v5.0.1` with the correct CSS build.

**Comment by angelikatyborska** [comment]
Note to self: fix heading with list, the list should not look like a list

File: test/bitstyles_phoenix/bitstyles_test.exs [hunk 1]
```diff
@@ -4,13 +4,84 @@ defmodule BitstylesPhoenix.BitstylesTest do
...
+      assert classname("u-padding-3xl\@l", "4.3.0") == "u-padding-xxxl\@l"
+      assert classname("u-margin-3xl-bottom", "4.3.0") == "u-margin-xxxl-bottom"
+      assert classname("u-margin-neg-3xl-bottom", "4.3.0") == "u-margin-neg-xxxl-bottom"
+      assert classname("u-list-none", "4.3.0") == "a-list-reset"
+      assert classname("a-badge--text", "4.3.0") == "a-badge--gray"
+      assert classname("u-gap-l", "4.3.0") == "u-gap-m"
+      assert classname("u-fg-text", "4.3.0") == "u-fg-gray-30"
+      assert classname("u-fg-text-darker", "4.3.0") == "u-fg-gray-50"
+      assert classname("u-bg-gray-darker", "4.3.0") == "u-bg-gray-80"
+      # TODO: top bottom
```
Comment (angelikatyborska) [inline test/bitstyles_phoenix/bitstyles_test.exs:79]:
```suggestion
```

File: lib/bitstyles_phoenix/bitstyles.ex [hunk 2]
```diff
@@ -14,12 +14,51 @@ defmodule BitstylesPhoenix.Bitstyles do
...
+    sizes_renaming = %{
+      "3xs" => "xxxs",
+      "2xs" => "xxs",
+      "2xl" => "xxl",
+      "3xl" => "xxxl"
+    }
+
+    class =
+      Enum.reduce(sizes_renaming, class, fn {new_size, old_size}, acc ->
+        String.replace(acc, "-#{new_size}", "-#{old_size}")
```
Comment (andreasknoepfle) [inline lib/bitstyles_phoenix/bitstyles.ex:36]: Shouldn't we at least make sure here, that the class starts with `u-` and is a bitstyles class?

File: lib/bitstyles_phoenix/component/dropdown.ex [hunk 3]
```diff
@@ -478,7 +514,16 @@ defmodule BitstylesPhoenix.Component.Dropdown do
     extra =
       assigns
       |> assigns_to_attributes()
-      |> Keyword.put_new(:variant, "menu")
+
+    extra =
+      if Bitstyles.version() >= "5.0.0" do
```
Comment (andreasknoepfle) [inline lib/bitstyles_phoenix/component/dropdown.ex:519]: We really have to start refactoring version comparisons to not use strings at some point 🙈 
But good for now.

**Review by andreasknoepfle — COMMENTED** [review]
Awesome work. 💚 I have small annotations on the version compatibility.

File: lib/bitstyles_phoenix/bitstyles.ex [hunk 4]
```diff
@@ -14,12 +14,51 @@ defmodule BitstylesPhoenix.Bitstyles do
...
+    class =
+      Enum.reduce(border_color_renaming, class, fn {new_border_color, old_border_color}, acc ->
+        String.replace(acc, new_border_color, old_border_color)
+      end)
+
+    class =
+      case class do
+        "u-list-none" -> "a-list-reset"
+        "a-badge--text" -> "a-badge--gray"
+        "u-gap-l" -> "u-gap-m"
```
[the remark this answers is not in the document]
Comment (angelikatyborska) [reply lib/bitstyles_phoenix/bitstyles.ex:53]: I'm not sure I understand...

> We have numerous places in the code in dima e.g. that use u-gap-l already. They will all be using u-gap-m afterwards if they pass through this.

Isn't this exactly what we want? The class `u-gap-m` is 4.3.0 has the same styles as `u-gap-l` in 5.0.0. The name of the class changed, but the exactly rem value used for the gap remained the same.

File: lib/bitstyles_phoenix/component/dropdown.ex (hunk 3 above)
Comment (angelikatyborska) [reply lib/bitstyles_phoenix/component/dropdown.ex:519]: Yes 😇 I'll try to squeeze in that work in my schedule after 5.0.0 and 6.0.0 upgrades are merged

File: lib/bitstyles_phoenix/bitstyles.ex (hunk 2 above)
Comment (angelikatyborska) [reply lib/bitstyles_phoenix/bitstyles.ex]: [`15d7c55` (#125)](https://github.com/bitcrowd/bitstyles_phoenix/pull/125/commits/15d7c55183b2924d2aa500b113b8aaf72092c99c)

File: lib/bitstyles_phoenix/bitstyles.ex (hunk 4 above)
[the remark this answers is not in the document]
Comment (andreasknoepfle) [reply lib/bitstyles_phoenix/bitstyles.ex:57]: But then are the values for `u-gap-m` and `u-gap-l` then same value by default now? 
What I am saying is also that we do not make an assumption on the "migratedness" of foreign code that uses the helpers. With this setting you basically can't use `u-gap-l` anymore with bistyles `4.3.0` even if you want to have that value if you have classes going through this helper.

[the remark this answers is not in the document]
Comment (andreasknoepfle) [reply lib/bitstyles_phoenix/bitstyles.ex:57]: Also keep in mind the values are always just values set by the sass variables, so one can always go back to the old values if they want to not migrate things.

[the remark this answers is not in the document]
Comment (angelikatyborska) [reply lib/bitstyles_phoenix/bitstyles.ex:57]: > But then are the values for u-gap-m and u-gap-l then same value by default now?

The class `u-gap-m` does not exist in the default configuration of bitstyles in 5.0.0, see: 

https://github.com/bitcrowd/bitstyles/blob/v5.0.1/scss/bitstyles/utilities/gap/_settings.scss#L4-L7

> What I am saying is also that we do not make an assumption on the "migratedness" of foreign code that uses the helpers. With this setting you basically can't use u-gap-l anymore with bistyles 4.3.0 even if you want to have that value if you have classes going through this helper.

I think I understand. It's fine to define class mappings in this module if the new class only exists in the new version, and the old class only exists in the old version. But here we have a new class that exists in both new and old version, but with a different meaning in the old version (`u-gap-l` from 4.3.0 would be equivalent to `u-gap-xl` in 5.0.0)

[the remark this answers is not in the document]
Comment (angelikatyborska) [reply lib/bitstyles_phoenix/bitstyles.ex]: [`c9e1f48` (#125)](https://github.com/bitcrowd/bitstyles_phoenix/pull/125/commits/c9e1f4842c2e5caf68911b17fee13856dd91fdc4)

**Review by andreasknoepfle — APPROVED** [review]
Fine from my perspective. Can you also prepare a `CHANGELOG` entry, listing the `u-gap-m` to `u-gap-l` change on the affected components.

**Pull request merged**